### DISCUSSION TITLE
Transforming differently when using Eloquent casting

### DISCUSSION THREAD

**Discussion by innocenzi** [spatie/laravel-data#579] [Ideas]
I often encounter situations where I want to have the current behavior for saving a data object to the database, but I want to have a different structure when transforming for the front-end.

My top example is enums. I want to save the backed enum as its int or string value, but I also need it to be workable on the front-end when using Inertia or Hybridly.

```php
final class PilotAssignmentData extends Data
{
    public function __construct(
        public readonly PilotGrade $grade,
        public readonly AircraftType $aircraft_type,
    ) {
    }
}
```

The above is an example in my application, where this `PilotAssignmentData` is actually saved as a collection. Another data object has a `DataCollectionOf(PilotAssignmentData::class)` property. 

I'm specifying this because it would be bothersome to have to duplicate a nested data object structure just for front-end usage, and I also can't add properties to `PilotAssignmentData` because they would be saved to the database.

What I want on the front-end is to obtain this data:

```json
"assignments": [
  {
    "grade": { "value": "captain", "human_readable": "Captain" },
    "aircraft_type": { "value": "PC12", "human_readable": "PC-12" }
  },
  {
    "grade": { "value": "first_officer", "human_readable": "First officer" },
    "aircraft_type": { "value": "PC12", "human_readable": "PC-12" }
  },
]
```

However, I still want it saved properly in the database: 
```json
"assignments": [
  { "grade": "captain", "aircraft_type": "PC12" },
  { "grade": "first_officer", "aircraft_type": "PC12" },
]
```

The `value` and `human_readable` parts could be returned from something like a transformer that I would make myself. The issue is to be able to:
- Use this specific transformer at a specific time (in my controller code for instance)
- Have corresponding TypeScript types generated, instead of the types for the normal transform

I don't think this is doable right now. I would be willing to implement the feature, but I think this requires a bit of brainstorming. Thoughts @rubenvanassche?

**Comment by rubenvanassche** [comment]
Yeah I think @sebastiandedeyne also already had this problem.

The thing is, right now we don't make a distinction between transforming for database or response.  I suggest we'll keep the transformation logic as is. This keeps the typescript definitions being generated correctly.

Also most of the times within your controllers you do this:

```
public function __invoke()
{
   return SomeData::from($something);
}
```

We actually have no idea what's happening here, is it a controller or a value being transformed to be stored somewhere?

That being said, in the case of storing a data object within a database, we'll always have a data cast. So I think we can create a hook, like beforeSavingToDatabase where the transformed array would be passed in so that it can be changed for correctly storing it within the database.

What do you think? Always open for suggestions.

**Reply by innocenzi** [reply]
Regarding your `beforeSavingToDatabase` suggestion, what would my example data object look like? I think I wouldn't be able to use `PilotGrade` as the type, right? Or would a transform be needed?

---

A alternative solution would be a `#[NonSerializable]` attribute that we could add to a property of a data object. In my example above, that would look like this:

```php
final class PilotAssignmentData extends Data
{
    public function __construct(
        public readonly PilotGrade $grade,
        public readonly AircraftType $aircraft_type,
        #[NonSerializable]
        public readonly array $grade_options,
        #[NonSerializable]
        public readonly array $aircraft_type_options,
    ) {
    }
}
```

Pros:
- Probably easy to implement (serialization-wide)
- Doesn't mess with TypeScript type generation

Cons: 
- Not sure how that would mess with a `::from` call where `grade_options` and `aircraft_type_options` are not passed (since those should be computed)
- Kind of irks me, looks like duplication of data, takes more bandwidth
- Not sure where the `grade` -> `grade_options` logic should be (an interface that the class implements?)

**Reply by innocenzi** [reply]
Basically, what I need is the ability to write a transformer for `laravel-data` and `typescript-transformer`, but to _not_ use that transformer when saving to the database. I think that formulated like this, this makes more sense.

**Reply by innocenzi** [reply]
I just took a stab at implementing that, but the fact that a type can only have _one_ transformer makes it non-viable to have a kind of transformer that can be ignored when serializing for the database...

So we would need to allow using multiple transformers, and use the first one that doesn't return `null` or something like that.

Another alternative seems to be sharing the context of the transform (eg. when transforming for Eloquent) and overriding individuals transformers, but that seems to be impossible to implement properly in a backwards-compatible way.

**Reply by innocenzi** [reply]
I tried the first option in https://github.com/spatie/laravel-data/pull/588, but testing in my own app, I have issues with casting (in the pipeline) when using in `DataCollection`. 

This comforts me in the idea that there should definitely be a way to transform only for the front-end, instead of a way to transform only for Eloquent casting, because data objects can be used elsewhere and we probably don't want the custom transform in those places (in my case, in a `DataCollection`).

**Reply by innocenzi** [reply]
~~I'm trying the second option, but I'm not succeeding; transforms seem to happen before I can give some context to the transformer. There's something that calls `toArray` before I can manually call `transform`, and I need to find out where.~~ 

**EDIT:** found where it was blocking, here's the PR: https://github.com/spatie/laravel-data/pull/589

However, this does seem like a good solution if we manage that. I can properly type and transform my objects/enums by extending the configs of `laravel-data` and `typescript-transformer`:

<details>
<summary>Details</summary>

As an example, the following transforms both the data and its types to what I want on the front-end:

```php
// Interface for my enums
interface HasOptions
{
    public function toOption(): array;
}

// For TypeScript Transformer
final class OptionsTransformer extends EnumTransformer
{
    protected function toEnumValue(ReflectionEnumBackedCase $case): string
    {
        if ($case->getEnum()->implementsInterface(HasOptions::class)) {
            return json_encode($case->getValue()->toOption());
        }

        $value = $case->getBackingValue();

        return \is_string($value) ? "'{$value}'" : "{$value}";
    }
}

// For Laravel Data
final class OptionsTransformer implements Transformer
{
    public function transform(DataProperty $property, mixed $value, ?string $context = null): null|string|array
    {
        if (!$value instanceof BackedEnum) {
            return null;
        }

        if ($context !== 'eloquent' && $value instanceof HasOptions) {
            return $value->toOption();
        }

        return $value->value;
    }
}
```

</details>
